openfl-tutorials/interactive_api/HPU/PyTorch_MedMNIST_2D/workspace/HPU_Pytorch_MedMNIST_2D.ipynb

```diff
@@ -0,0 +1,627 @@
+   "id": "4ff463bd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "TI = TaskInterface()\n",
+    "\n",
+    "train_custom_params={'criterion':criterion,'task':task, 'device':'hpu'}\n",
+    "\n",
+    "# Task interface currently supports only standalone functions.\n",
+    "@TI.add_kwargs(**train_custom_params)\n",
+    "@TI.register_fl_task(model='model', data_loader='train_loader',\n",
+    "                     device='device', optimizer='optimizer')\n",
+    "def train(model, train_loader, device, optimizer, criterion, task):\n",
+    "    total_loss = []\n",
+    "    \n",
+    "    train_loader = tqdm.tqdm(train_loader, desc=\"train\")\n",
+    "    model.train()\n",
+    "    model.to(device)\n",
+    "    \n",
+    "    for inputs, targets in train_loader:\n",
+    "    \n",
+    "        optimizer.zero_grad()\n",
+    "        outputs = model(inputs.to(device))\n",
+    "        \n",
+    "        if task == 'multi-label, binary-class':\n",
+    "            targets = targets.to(torch.float32).to(device)\n",
+    "            loss = criterion(outputs, targets)\n",
+    "        else:\n",
+    "            targets = torch.squeeze(targets, 1).long().to(device)\n",
+    "            loss = criterion(outputs, targets)\n",
+    "        \n",
+    "        total_loss.append(loss.item())\n",
+    "        \n",
+    "        loss.backward()\n",
+    "        htcore.mark_step()\n",
+    "        optimizer.step()\n",
+    "        htcore.mark_step()\n",
+    "\n",
+    "    return {'train_loss': np.mean(total_loss),}\n",
+    "\n",
+    "\n",
+    "\n",
+    "val_custom_params={'criterion':criterion, 'task':task, 'device':'hpu'}\n",
+    "\n",
+    "@TI.add_kwargs(**val_custom_params)\n",
+    "@TI.register_fl_task(model='model', data_loader='val_loader', device='device')\n",
+    "def validate(model, val_loader, device, criterion, task):\n",
+    "\n",
+    "    val_loader = tqdm.tqdm(val_loader, desc=\"validate\")\n",
+    "    model.eval()\n",
+    "    model.to(device)\n",
+    "\n",
+    "    val_score = 0\n",
+    "    total_samples = 0\n",
+    "    total_loss = []\n",
+    "    y_score = torch.tensor([]).to(device)\n",
+    "\n",
+    "    with torch.no_grad():\n",
+    "        for inputs, targets in val_loader:\n",
+    "            outputs = model(inputs.to(device))\n",
+    "            \n",
+    "            if task == 'multi-label, binary-class':\n",
+    "                targets = targets.to(torch.float32).to(device)\n",
+    "                loss = criterion(outputs, targets)\n",
+    "                m = nn.Sigmoid()\n",
+    "                outputs = m(outputs).to(device)\n",
+    "            else:\n",
+    "                targets = torch.squeeze(targets, 1).long().to(device)\n",
+    "                loss = criterion(outputs, targets) \n",
+    "                m = nn.Softmax(dim=1)\n",
+    "                outputs = m(outputs).to(device)\n",
+    "                targets = targets.to(torch.float32)\n",
```

Comment:
```python
targets = targets.to(torch.float32)
# targets = targets.float().resize_(len(targets), 1)
# right way to use float() is tensor.to(torch.float32) else gives RuntimeError: expected scalar type Long but found Float
# The resize_(len(targets), 1) operation in above line causes below graph compile error, hence removed it
# RuntimeError: FATAL ERROR :: MODULE:BRIDGE syn compile encountered : Graph compile failed
```
1. Why does it need to be cast to float? AFAIK, `nn.CrossentropyLoss()` works on long tensors.
2. From the comment block, it is not clear what `resize_` is supposed to do, other than the fact that it throws an exception. `resize_` manipulates tensor shape. Alternatively, if `resize_` is not required, which other function is taking its role?

Overall, I think both the code and the comment has the potential to create unnecessary confusion. Generally, a combination of below statements is used without any explicit typecasts:
```python
inputs = inputs.to(device)  # inputs = [bs, H, W, C] or [bs, C, H, W] both are equivalent
targets = targets.to(device)  # targets = [bs] a.k.a 1D array of long integers

# Forward pass
optimizer.zero_grad()
outputs = model(inputs)  # outputs = [bs, num_classes]
pred = outputs.argmax(1)  # pred = [bs] of long integers, same as targets
loss = criterion(outputs, targets)  # loss = [bs] of floats

# Backward and step
loss.backward()
optimizer.step()  # And habana-specific mitigation if any.

# Log
acc += pred.eq(targets).sum().cpu().numpy()
```

Reply:
Hi @shreyasnc1, I also had this query, casting to float and resize_() are not done in training part, it can be seen only in validation. These two operations do not have any impact on the accuracy. Do we need to keep this?

The comment was added just to bring it to the notice that these lines were part of original code and that resize if kept is causing that error

Reply:
`reshape` or `view` are a better alternative. 
Please check for the correctness of the train and validation loop by comparing previous with the modified/removed `resize_` or `squeeze` operations.

Reply:
Yes, view is the better option. I have tried with it, but we do not need these operations because of the reasons mentioned above. I have validated the the example after removing it. It is working fine.